### PR TITLE
process replay: show added and removed messages

### LINKED ISSUE #25463: process replay improvements/fixes
- [x] Print additions or removals when "logs are not same length" ([an example](https://github.com/commaai/openpilot/runs/7852984165?check_suite_focus=true)) - https://github.com/commaai/openpilot/pull/30173
- [x] Fix random failures (process replay can fail with above message, then pass when run right after)
- [x] Random regen failures (this line can fail for driver monitoring: https://github.com/commaai/openpilot/blob/master/selfdrive/modeld/runners/onnx_runner.py#L18. Traced it back to a crashing or hanging memcpy in `dmonitoring.cc`)
- [ ] print address for diff CAN messages

### PR DESCRIPTION
**Pull request by sshane**
Split from https://github.com/commaai/openpilot/pull/30136 so we can show new/removed events -- very difficult to debug changes without this (have to write a script every time or open up PJ and compare for the entire segment)

If length of logs don't match, it will switch to printing new or removed messages instead of exception, else same dictdiffer logic

TODO:
- [x] simplify and clean up
- [x] keys are missing?

An example:

```python
***** tested against commit 7553d9123f418fd81ef811df4e130ab153ac489a *****
***** differences for segment aregen765AF3D2CB5|2023-05-10--14-54-23--0 *****
*** process: controlsd ***
	ref: https://commadataci.blob.core.windows.net/openpilotci/aregen765AF3D2CB5|2023-05-10--14-54-23--0_controlsd_7553d9123f418fd81ef811df4e130ab153ac489a.bz2
	new: /home/batman/openpilot/openpilot/selfdrive/test/process_replay/fakedata/aregen765AF3D2CB5|2023-05-10--14-54-23--0_controlsd_7f1baf85c97a965960821dfabf1306096c6ab40c.bz2

	['remove', '', [('carEvents', [{'name': 'pedalPressed', 'enable': False, 'noEntry': True, 'warning': False, 'userDisable': True, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False}, {'name': 'gasPressedOverride', 'enable': False, 'noEntry': False, 'warning': False, 'userDisable': False, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': True, 'overrideLateral': False}, {'name': 'driverUnresponsive', 'enable': False, 'noEntry': False, 'warning': True, 'userDisable': False, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False}])]]
	['remove', '', [('carEvents', [{'name': 'pedalPressed', 'enable': False, 'noEntry': True, 'warning': False, 'userDisable': True, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False}, {'name': 'gasPressedOverride', 'enable': False, 'noEntry': False, 'warning': False, 'userDisable': False, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': True, 'overrideLateral': False}, {'name': 'pcmDisable', 'enable': False, 'noEntry': False, 'warning': False, 'userDisable': True, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False}])]]
	['remove', '', [('carEvents', [{'name': 'pedalPressed', 'enable': False, 'noEntry': True, 'warning': False, 'userDisable': True, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False}, {'name': 'gasPressedOverride', 'enable': False, 'noEntry': False, 'warning': False, 'userDisable': False, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': True, 'overrideLateral': False}, {'name': 'pcmDisable', 'enable': False, 'noEntry': False, 'warning': False, 'userDisable': True, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False}])]]
	['add', '', [('carEvents', [{'name': 'radarFault', 'enable': False, 'noEntry': True, 'warning': False, 'userDisable': False, 'softDisable': True, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False}])]]
```

### CODE REVIEW

**Comment by sshane** [comment]
Some changes:

- Logs are not expected to be misaligned (type of event/ie. `msg.which()`), so that check is removed. 
  - So if length of logs for each service type is different, then we show the added or removed messages using `dict.__eq__`
- diff.txt now shows diff for each service [in chunks](https://github.com/commaai/openpilot/assets/25857203/e0444df3-38e9-49f7-b7f3-87b353ace6f7), instead of all messages image.

**Comment by sshane** [comment]
Diff example removing gas press as a way to disengage on our Prius segment:

<details>
<summary>Before</summary>

***This is the summary from process replay since diff.txt shows no diff:**
```python
***** results for segment aregen8D6A8B36E8D|2023-05-10--14-32-38--0 *****
    controlsd
        ref: https://commadataci.blob.core.windows.net/openpilotci/aregen8D6A8B36E8D|2023-05-10--14-32-38--0_controlsd_7553d9123f418fd81ef811df4e130ab153ac489a.bz2
        new: /home/batman/openpilot/openpilot/selfdrive/test/process_replay/fakedata/aregen8D6A8B36E8D|2023-05-10--14-32-38--0_controlsd_f4b2b70791144a82d9c9a215996a5f51038a552d.bz2

        logs are not same length: 24095 VS 24094
                Counter({'controlsState': 6038, 'carState': 6038, 'carControl': 6038, 'sendcan': 5834, 'carEvents': 145, 'carParams': 2})
                Counter({'controlsState': 6038, 'carState': 6038, 'carControl': 6038, 'sendcan': 5834, 'carEvents': 144, 'carParams': 2})

TEST FAILED
```

</details>

<details>
<summary>After</summary>

```python
***** tested against commit 7553d9123f418fd81ef811df4e130ab153ac489a *****
***** differences for segment aregen8D6A8B36E8D|2023-05-10--14-32-38--0 *****
*** process: controlsd ***
	ref: https://commadataci.blob.core.windows.net/openpilotci/aregen8D6A8B36E8D|2023-05-10--14-32-38--0_controlsd_7553d9123f418fd81ef811df4e130ab153ac489a.bz2
	new: /home/batman/openpilot/openpilot/selfdrive/test/process_replay/fakedata/aregen8D6A8B36E8D|2023-05-10--14-32-38--0_controlsd_d3a638f1275c9a9acf523fdd099928f3486ca245.bz2

	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', 'controlsState.alertType', ('pedalPressed/userDisable', 'wrongCarMode/userDisable'))
	('change', ['carState', 'events', 0, 'name'], ('pedalPressed', 'gasPressedOverride'))
	('change', ['carState', 'events', 0, 'noEntry'], (True, False))
	('change', ['carState', 'events', 0, 'userDisable'], (True, False))
	('change', ['carState', 'events', 0, 'overrideLongitudinal'], (False, True))
	('change', ['carState', 'events', 1, 'name'], ('gasPressedOverride', 'wrongCarMode'))
	('change', ['carState', 'events', 1, 'noEntry'], (False, True))
	('change', ['carState', 'events', 1, 'userDisable'], (False, True))
	('change', ['carState', 'events', 1, 'overrideLongitudinal'], (True, False))
	('change', ['carState', 'events', 2, 'name'], ('wrongCarMode', 'pcmDisable'))
	('change', ['carState', 'events', 2, 'noEntry'], (True, False))
	('remove', 'carState.events', [(3, {'name': 'pcmDisable', 'enable': False, 'noEntry': False, 'warning': False, 'userDisable': True, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False})])
	('change', [143, 'carEvents', 0, 'name'], ('pedalPressed', 'gasPressedOverride'))
	('change', [143, 'carEvents', 0, 'noEntry'], (True, False))
	('change', [143, 'carEvents', 0, 'userDisable'], (True, False))
	('change', [143, 'carEvents', 0, 'overrideLongitudinal'], (False, True))
	('change', [143, 'carEvents', 1, 'name'], ('gasPressedOverride', 'wrongCarMode'))
	('change', [143, 'carEvents', 1, 'noEntry'], (False, True))
	('change', [143, 'carEvents', 1, 'userDisable'], (False, True))
	('change', [143, 'carEvents', 1, 'overrideLongitudinal'], (True, False))
	('change', [143, 'carEvents', 2, 'name'], ('wrongCarMode', 'pcmDisable'))
	('change', [143, 'carEvents', 2, 'noEntry'], (True, False))
	('remove', [143, 'carEvents'], [(3, {'name': 'pcmDisable', 'enable': False, 'noEntry': False, 'warning': False, 'userDisable': True, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False})])
	('remove', '', [(144, {'carEvents': [{'name': 'gasPressedOverride', 'enable': False, 'noEntry': False, 'warning': False, 'userDisable': False, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': True, 'overrideLateral': False}, {'name': 'wrongCarMode', 'enable': False, 'noEntry': True, 'warning': False, 'userDisable': True, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False}, {'name': 'pcmDisable', 'enable': False, 'noEntry': False, 'warning': False, 'userDisable': True, 'softDisable': False, 'immediateDisable': False, 'preEnable': False, 'permanent': False, 'overrideLongitudinal': False, 'overrideLateral': False}], 'logMonoTime': 0, 'valid': False})])
```
</details>

File: selfdrive/test/process_replay/compare_logs.py [hunk 1]
```diff
@@ -52,41 +52,49 @@ def compare_logs(log1, log2, ignore_fields=None, ignore_msgs=None, tolerance=Non
     for log in (log1, log2)
   )
 
-  if len(log1) != len(log2):
-    cnt1 = Counter(m.which() for m in log1)
-    cnt2 = Counter(m.which() for m in log2)
-    raise Exception(f"logs are not same length: {len(log1)} VS {len(log2)}\n\t\t{cnt1}\n\t\t{cnt2}")
+  msgs_by_which = defaultdict(lambda: defaultdict(list))
```
Comment (sshane) [inline selfdrive/test/process_replay/compare_logs.py:55]:
```suggestion
  msgs_by_service = defaultdict(lambda: defaultdict(list))
```

**Comment by fredyshox** [comment]
Cases this might now catch anymore:
- messages are the same as for content, but couple of messages of different type are not sent in the same order. this is currently caught by message alignment check

**Comment by sshane** [comment]
Test which fails on this branch: https://github.com/commaai/openpilot/pull/30197

**Comment by sshane** [comment]
going to close for now since you can run `./test_processes.py --ignore-msgs carEvents`, but should re-visit if another service commonly adds/removes messages (one of the GPS daemons I think?)

**Pull request closed without merging**